pkg/controllers/provisioning/v1alpha1/reallocator/controller.go

```diff
@@ -13,3 +13,67 @@ limitations under the License.
 */
 
 package reallocator
+
+import (
+	"context"
+	"fmt"
+	"github.com/awslabs/karpenter/pkg/apis/provisioning/v1alpha1"
+	"github.com/awslabs/karpenter/pkg/cloudprovider"
+	"github.com/awslabs/karpenter/pkg/controllers"
+	"go.uber.org/zap"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"time"
+)
+
+// Controller for the resource
+type Controller struct {
+	filter        *Filter
+	cloudProvider cloudprovider.Factory
+}
+
+// For returns the resource this controller is for.
+func (c *Controller) For() controllers.Object {
+	return &v1alpha1.Provisioner{}
+}
+
+// Owns returns the resources owned by this controller's resource.
+func (c *Controller) Owns() []controllers.Object {
+	return []controllers.Object{}
+}
+
+func (c *Controller) Interval() time.Duration {
+	return 5 * time.Second
+}
+
+// NewController constructs a controller instance
+func NewController(kubeClient client.Client, cloudProvider cloudprovider.Factory) *Controller {
+	return &Controller{
+		filter:        &Filter{kubeClient: kubeClient},
+		cloudProvider: cloudProvider,
+	}
+}
+
+// Reconcile executes an allocation control loop for the resource
+func (c *Controller) Reconcile(object controllers.Object) error {
+	ctx := context.TODO()
+
+	// 1. Filter all nodes with Karpenter TTL label
+
+	// 2. Remove TTL from nodes with TTL label that have pods and if past TTL, cordon/drain node
+
+	// 3. Filter under-utilized nodes
+	nodes, err := c.filter.GetUnderUtilizedNodes(ctx)
+	if err != nil {
+		return fmt.Errorf("filtering nodes, %w", err)
+	}
+	if len(nodes) == 0 {
+		return nil
+	}
+	zap.S().Infof("Found %d under-utilized nodes", len(nodes))
```

Comment:
nit: underutilize is one work

Reply:
huh didn't even know. Just assumed it would have a hiphen. Thanks